server/src/test/java/io/crate/integrationtests/JoinIntegrationTest.java

```diff
@@ -1335,15 +1335,15 @@ public void test_alias_in_left_join_condition() {
 
         execute("EXPLAIN (COSTS FALSE)" + stmt);
         assertThat(response).hasLines(
-                "Eval[id, reference]",
-                "  └ NestedLoopJoin[LEFT | ((cluster_id = id) AND (kind = 'bar'))]",
-                "    ├ HashJoin[(cluster_id = id)]",
-                "    │  ├ HashJoin[(subscription_id = id)]",
-                "    │  │  ├ Collect[doc.t3 | [id, reference] | (reference = 'bazinga')]",
-                "    │  │  └ Collect[doc.t1 | [subscription_id, id] | true]",
-                "    │  └ Collect[doc.t2 | [cluster_id] | (kind = 'bar')]",
-                "    └ Rename[cluster_id, kind] AS temp",
-                "      └ Collect[doc.t2 | [cluster_id, kind] | true]"
+            "Eval[id, reference]",
```

Comment:
This is a nice example why this change makes sense. The constant join condition from the left-join gets extracted and pushed down and the join can be converted to inner-join which leads to a more efficient query plan using solely hash-joins.

```
+------------------------------------------------------+----------------------------------------------------------------------------------------+
| STEP                                                 | QUERY PLAN                                                                             |
+------------------------------------------------------+----------------------------------------------------------------------------------------+
| Initial logical plan                                 | Eval[id, reference] (rows=unknown)                                                     |
|                                                      |   └ Join[LEFT | ((cluster_id = id) AND (kind = 'bar'))] (rows=unknown)                 |
|                                                      |     ├ Join[INNER | ((cluster_id = id) AND (kind = 'bar'))] (rows=unknown)              |
|                                                      |     │  ├ Filter[(reference = 'bazinga')] (rows=0)                                      |
|                                                      |     │  │  └ Join[INNER | (subscription_id = id)] (rows=unknown)                        |
|                                                      |     │  │    ├ Collect[doc.t3 | [id, reference] | true] (rows=unknown)                  |
|                                                      |     │  │    └ Collect[doc.t1 | [subscription_id, id] | true] (rows=unknown)            |
|                                                      |     │  └ Collect[doc.t2 | [cluster_id, kind] | true] (rows=unknown)                    |
|                                                      |     └ Rename[cluster_id, kind] AS temp (rows=unknown)                                  |
|                                                      |       └ Collect[doc.t2 | [cluster_id, kind] | true] (rows=unknown)                     |
| optimizer_extract_constant_join_conditions_to_filter | Eval[id, reference] (rows=0)                                                           |
|                                                      |   └ Filter[(kind = 'bar')] (rows=0)                                                    |
|                                                      |     └ Join[LEFT | (cluster_id = id)] (rows=unknown)                                    |
|                                                      |       ├ Join[INNER | ((cluster_id = id) AND (kind = 'bar'))] (rows=unknown)            |
|                                                      |       │  ├ Filter[(reference = 'bazinga')] (rows=0)                                    |
|                                                      |       │  │  └ Join[INNER | (subscription_id = id)] (rows=unknown)                      |
|                                                      |       │  │    ├ Collect[doc.t3 | [id, reference] | true] (rows=unknown)                |
|                                                      |       │  │    └ Collect[doc.t1 | [subscription_id, id] | true] (rows=unknown)          |
|                                                      |       │  └ Collect[doc.t2 | [cluster_id, kind] | true] (rows=unknown)                  |
|                                                      |       └ Rename[cluster_id, kind] AS temp (rows=unknown)                                |
|                                                      |         └ Collect[doc.t2 | [cluster_id, kind] | true] (rows=unknown)                   |
| optimizer_rewrite_filter_on_outer_join_to_inner_join | Eval[id, reference] (rows=unknown)                                                     |
|                                                      |   └ Join[INNER | (cluster_id = id)] (rows=unknown)                                     |
|                                                      |     ├ Join[INNER | ((cluster_id = id) AND (kind = 'bar'))] (rows=unknown)              |
|                                                      |     │  ├ Filter[(reference = 'bazinga')] (rows=0)                                      |
|                                                      |     │  │  └ Join[INNER | (subscription_id = id)] (rows=unknown)                        |
|                                                      |     │  │    ├ Collect[doc.t3 | [id, reference] | true] (rows=unknown)                  |
|                                                      |     │  │    └ Collect[doc.t1 | [subscription_id, id] | true] (rows=unknown)            |
|                                                      |     │  └ Collect[doc.t2 | [cluster_id, kind] | true] (rows=unknown)                    |
|                                                      |     └ Filter[(kind = 'bar')] (rows=0)                                                  |
|                                                      |       └ Rename[cluster_id, kind] AS temp (rows=unknown)                                |
|                                                      |         └ Collect[doc.t2 | [cluster_id, kind] | true] (rows=unknown)                   |
| optimizer_rewrite_join_plan                          | Eval[id, reference] (rows=unknown)                                                     |
|                                                      |   └ HashJoin[(cluster_id = id)] (rows=unknown)                                         |
|                                                      |     ├ Join[INNER | ((cluster_id = id) AND (kind = 'bar'))] (rows=unknown)              |
|                                                      |     │  ├ Filter[(reference = 'bazinga')] (rows=0)                                      |
|                                                      |     │  │  └ Join[INNER | (subscription_id = id)] (rows=unknown)                        |
|                                                      |     │  │    ├ Collect[doc.t3 | [id, reference] | true] (rows=unknown)                  |
|                                                      |     │  │    └ Collect[doc.t1 | [subscription_id, id] | true] (rows=unknown)            |
|                                                      |     │  └ Collect[doc.t2 | [cluster_id, kind] | true] (rows=unknown)                    |
|                                                      |     └ Filter[(kind = 'bar')] (rows=0)                                                  |
|                                                      |       └ Rename[cluster_id, kind] AS temp (rows=unknown)                                |
|                                                      |         └ Collect[doc.t2 | [cluster_id, kind] | true] (rows=unknown)                   |
| optimizer_extract_constant_join_conditions_to_filter | Eval[id, reference] (rows=unknown)                                                     |
|                                                      |   └ HashJoin[(cluster_id = id)] (rows=unknown)                                         |
|                                                      |     ├ Filter[(kind = 'bar')] (rows=0)                                                  |
|                                                      |     │  └ Join[INNER | (cluster_id = id)] (rows=unknown)                                |
|                                                      |     │    ├ Filter[(reference = 'bazinga')] (rows=0)                                    |
|                                                      |     │    │  └ Join[INNER | (subscription_id = id)] (rows=unknown)                      |
|                                                      |     │    │    ├ Collect[doc.t3 | [id, reference] | true] (rows=unknown)                |
|                                                      |     │    │    └ Collect[doc.t1 | [subscription_id, id] | true] (rows=unknown)          |
|                                                      |     │    └ Collect[doc.t2 | [cluster_id, kind] | true] (rows=unknown)                  |
|                                                      |     └ Filter[(kind = 'bar')] (rows=0)                                                  |
|                                                      |       └ Rename[cluster_id, kind] AS temp (rows=unknown)                                |
|                                                      |         └ Collect[doc.t2 | [cluster_id, kind] | true] (rows=unknown)                   |
| optimizer_move_filter_beneath_join                   | Eval[id, reference] (rows=unknown)                                                     |
|                                                      |   └ HashJoin[(cluster_id = id)] (rows=unknown)                                         |
|                                                      |     ├ Join[INNER | (cluster_id = id)] (rows=unknown)                                   |
|                                                      |     │  ├ Filter[(reference = 'bazinga')] (rows=0)                                      |
|                                                      |     │  │  └ Join[INNER | (subscription_id = id)] (rows=unknown)                        |
|                                                      |     │  │    ├ Collect[doc.t3 | [id, reference] | true] (rows=unknown)                  |
|                                                      |     │  │    └ Collect[doc.t1 | [subscription_id, id] | true] (rows=unknown)            |
|                                                      |     │  └ Filter[(kind = 'bar')] (rows=0)                                               |
|                                                      |     │    └ Collect[doc.t2 | [cluster_id, kind] | true] (rows=unknown)                  |
|                                                      |     └ Filter[(kind = 'bar')] (rows=0)                                                  |
|                                                      |       └ Rename[cluster_id, kind] AS temp (rows=unknown)                                |
|                                                      |         └ Collect[doc.t2 | [cluster_id, kind] | true] (rows=unknown)                   |
| optimizer_rewrite_join_plan                          | Eval[id, reference] (rows=unknown)                                                     |
|                                                      |   └ HashJoin[(cluster_id = id)] (rows=unknown)                                         |
|                                                      |     ├ HashJoin[(cluster_id = id)] (rows=unknown)                                       |
|                                                      |     │  ├ Filter[(reference = 'bazinga')] (rows=0)                                      |
|                                                      |     │  │  └ Join[INNER | (subscription_id = id)] (rows=unknown)                        |
|                                                      |     │  │    ├ Collect[doc.t3 | [id, reference] | true] (rows=unknown)                  |
|                                                      |     │  │    └ Collect[doc.t1 | [subscription_id, id] | true] (rows=unknown)            |
|                                                      |     │  └ Filter[(kind = 'bar')] (rows=0)                                               |
|                                                      |     │    └ Collect[doc.t2 | [cluster_id, kind] | true] (rows=unknown)                  |
|                                                      |     └ Filter[(kind = 'bar')] (rows=0)                                                  |
|                                                      |       └ Rename[cluster_id, kind] AS temp (rows=unknown)                                |
|                                                      |         └ Collect[doc.t2 | [cluster_id, kind] | true] (rows=unknown)                   |
| optimizer_move_filter_beneath_join                   | Eval[id, reference] (rows=unknown)                                                     |
|                                                      |   └ HashJoin[(cluster_id = id)] (rows=unknown)                                         |
|                                                      |     ├ HashJoin[(cluster_id = id)] (rows=unknown)                                       |
|                                                      |     │  ├ Join[INNER | (subscription_id = id)] (rows=unknown)                           |
|                                                      |     │  │  ├ Filter[(reference = 'bazinga')] (rows=0)                                   |
|                                                      |     │  │  │  └ Collect[doc.t3 | [id, reference] | true] (rows=unknown)                 |
|                                                      |     │  │  └ Collect[doc.t1 | [subscription_id, id] | true] (rows=unknown)              |
|                                                      |     │  └ Filter[(kind = 'bar')] (rows=0)                                               |
|                                                      |     │    └ Collect[doc.t2 | [cluster_id, kind] | true] (rows=unknown)                  |
|                                                      |     └ Filter[(kind = 'bar')] (rows=0)                                                  |
|                                                      |       └ Rename[cluster_id, kind] AS temp (rows=unknown)                                |
|                                                      |         └ Collect[doc.t2 | [cluster_id, kind] | true] (rows=unknown)                   |
| optimizer_rewrite_join_plan                          | Eval[id, reference] (rows=unknown)                                                     |
|                                                      |   └ HashJoin[(cluster_id = id)] (rows=unknown)                                         |
|                                                      |     ├ HashJoin[(cluster_id = id)] (rows=unknown)                                       |
|                                                      |     │  ├ HashJoin[(subscription_id = id)] (rows=unknown)                               |
|                                                      |     │  │  ├ Filter[(reference = 'bazinga')] (rows=0)                                   |
|                                                      |     │  │  │  └ Collect[doc.t3 | [id, reference] | true] (rows=unknown)                 |
|                                                      |     │  │  └ Collect[doc.t1 | [subscription_id, id] | true] (rows=unknown)              |
|                                                      |     │  └ Filter[(kind = 'bar')] (rows=0)                                               |
|                                                      |     │    └ Collect[doc.t2 | [cluster_id, kind] | true] (rows=unknown)                  |
|                                                      |     └ Filter[(kind = 'bar')] (rows=0)                                                  |
|                                                      |       └ Rename[cluster_id, kind] AS temp (rows=unknown)                                |
|                                                      |         └ Collect[doc.t2 | [cluster_id, kind] | true] (rows=unknown)                   |
| optimizer_merge_filter_and_collect                   | Eval[id, reference] (rows=unknown)                                                     |
|                                                      |   └ HashJoin[(cluster_id = id)] (rows=unknown)                                         |
|                                                      |     ├ HashJoin[(cluster_id = id)] (rows=unknown)                                       |
|                                                      |     │  ├ HashJoin[(subscription_id = id)] (rows=unknown)                               |
|                                                      |     │  │  ├ Collect[doc.t3 | [id, reference] | (reference = 'bazinga')] (rows=unknown) |
|                                                      |     │  │  └ Collect[doc.t1 | [subscription_id, id] | true] (rows=unknown)              |
|                                                      |     │  └ Filter[(kind = 'bar')] (rows=0)                                               |
|                                                      |     │    └ Collect[doc.t2 | [cluster_id, kind] | true] (rows=unknown)                  |
|                                                      |     └ Filter[(kind = 'bar')] (rows=0)                                                  |
|                                                      |       └ Rename[cluster_id, kind] AS temp (rows=unknown)                                |
|                                                      |         └ Collect[doc.t2 | [cluster_id, kind] | true] (rows=unknown)                   |
| optimizer_merge_filter_and_collect                   | Eval[id, reference] (rows=unknown)                                                     |
|                                                      |   └ HashJoin[(cluster_id = id)] (rows=unknown)                                         |
|                                                      |     ├ HashJoin[(cluster_id = id)] (rows=unknown)                                       |
|                                                      |     │  ├ HashJoin[(subscription_id = id)] (rows=unknown)                               |
|                                                      |     │  │  ├ Collect[doc.t3 | [id, reference] | (reference = 'bazinga')] (rows=unknown) |
|                                                      |     │  │  └ Collect[doc.t1 | [subscription_id, id] | true] (rows=unknown)              |
|                                                      |     │  └ Collect[doc.t2 | [cluster_id, kind] | (kind = 'bar')] (rows=unknown)          |
|                                                      |     └ Filter[(kind = 'bar')] (rows=0)                                                  |
|                                                      |       └ Rename[cluster_id, kind] AS temp (rows=unknown)                                |
|                                                      |         └ Collect[doc.t2 | [cluster_id, kind] | true] (rows=unknown)                   |
| optimizer_move_filter_beneath_rename                 | Eval[id, reference] (rows=unknown)                                                     |
|                                                      |   └ HashJoin[(cluster_id = id)] (rows=unknown)                                         |
|                                                      |     ├ HashJoin[(cluster_id = id)] (rows=unknown)                                       |
|                                                      |     │  ├ HashJoin[(subscription_id = id)] (rows=unknown)                               |
|                                                      |     │  │  ├ Collect[doc.t3 | [id, reference] | (reference = 'bazinga')] (rows=unknown) |
|                                                      |     │  │  └ Collect[doc.t1 | [subscription_id, id] | true] (rows=unknown)              |
|                                                      |     │  └ Collect[doc.t2 | [cluster_id, kind] | (kind = 'bar')] (rows=unknown)          |
|                                                      |     └ Rename[cluster_id, kind] AS temp (rows=0)                                        |
|                                                      |       └ Filter[(kind = 'bar')] (rows=0)                                                |
|                                                      |         └ Collect[doc.t2 | [cluster_id, kind] | true] (rows=unknown)                   |
| optimizer_merge_filter_and_collect                   | Eval[id, reference] (rows=unknown)                                                     |
|                                                      |   └ HashJoin[(cluster_id = id)] (rows=unknown)                                         |
|                                                      |     ├ HashJoin[(cluster_id = id)] (rows=unknown)                                       |
|                                                      |     │  ├ HashJoin[(subscription_id = id)] (rows=unknown)                               |
|                                                      |     │  │  ├ Collect[doc.t3 | [id, reference] | (reference = 'bazinga')] (rows=unknown) |
|                                                      |     │  │  └ Collect[doc.t1 | [subscription_id, id] | true] (rows=unknown)              |
|                                                      |     │  └ Collect[doc.t2 | [cluster_id, kind] | (kind = 'bar')] (rows=unknown)          |
|                                                      |     └ Rename[cluster_id, kind] AS temp (rows=unknown)                                  |
|                                                      |       └ Collect[doc.t2 | [cluster_id, kind] | (kind = 'bar')] (rows=unknown)           |
| Final logical plan                                   | Eval[id, reference] (rows=unknown)                                                     |
|                                                      |   └ HashJoin[(cluster_id = id)] (rows=unknown)                                         |
|                                                      |     ├ HashJoin[(cluster_id = id)] (rows=unknown)                                       |
|                                                      |     │  ├ HashJoin[(subscription_id = id)] (rows=unknown)                               |
|                                                      |     │  │  ├ Collect[doc.t3 | [id, reference] | (reference = 'bazinga')] (rows=unknown) |
|                                                      |     │  │  └ Collect[doc.t1 | [subscription_id, id] | true] (rows=unknown)              |
|                                                      |     │  └ Collect[doc.t2 | [cluster_id] | (kind = 'bar')] (rows=unknown)                |
|                                                      |     └ Rename[cluster_id] AS temp (rows=unknown)                                        |
|                                                      |       └ Collect[doc.t2 | [cluster_id] | (kind = 'bar')] (rows=unknown)                 |
+------------------------------------------------------+----------------------------------------------------------------------------------------+
EXPLAIN 14 rows in set (0.026 sec)
```